server/queue/queue-api/src/test/java/org/apache/james/queue/api/PriorityMailQueueContract.java

```diff
@@ -0,0 +1,225 @@
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME0)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 0)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME7)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 7)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME4)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 4)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME2)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 2)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME5)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 5)
+            .build());
+
+        ImmutableList<MailQueue.MailQueueItem> items = IntStream.range(1, 11).boxed()
+            .map(Throwing.function(i -> {
+                MailQueue.MailQueueItem item1 = getMailQueue().deQueue();
+                item1.done(true);
+                return item1;
+            }))
+            .collect(Guavate.toImmutableList());
+
+        assertThat(items)
+            .extracting(MailQueue.MailQueueItem::getMail)
+            .extracting(Mail::getName)
+            .containsExactly(NAME9, NAME8, NAME7, NAME6, NAME5, NAME4, NAME3, NAME2, NAME1, NAME0);
+    }
+
+    @Test
+    default void negativePriorityShouldBeConsideredZero() throws Exception {
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME2)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, -1)
+            .build());
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME1)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 1)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo(NAME1);
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo(NAME2);
+    }
+
+    @Test
+    default void tooBigPriorityShouldBeConsideredMaximum() throws Exception {
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME2)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 12)
+            .build());
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME1)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 8)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo(NAME2);
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo(NAME1);
+    }
```

Comment:
you could also check that a mail with priority 12 is not before a mail with priority 10

Reply:
Maximum priority is 9, not 10.

Also, I believe I messed up with priority edge-case testing as reported by @mbaechler . I will have a finner look on it.